astronomy/orbital_elements_test.cpp

```diff
@@ -0,0 +1,422 @@
+      IsNear(0.92 * Micro(Second)));
+
+  EXPECT_THAT(elements.nodal_precession(), IsNear(1.0 * Degree / JulianYear));
+
+  // Mean element values.
+  EXPECT_THAT(AbsoluteError(*initial_osculating.semimajor_axis,
+                            elements.mean_semimajor_axis_range().midpoint()),
+              IsNear(100 * Micro(Metre)));
+  EXPECT_THAT(AbsoluteError(*initial_osculating.eccentricity,
+                            elements.mean_eccentricity_range().midpoint()),
+              IsNear(1.5e-11));
+  EXPECT_THAT(AbsoluteError(initial_osculating.inclination,
+                            elements.mean_inclination_range().midpoint()),
+              IsNear(0.56 * Micro(ArcSecond)));
+  EXPECT_THAT(AbsoluteError(
+                  initial_osculating.longitude_of_ascending_node,
+                  elements.mean_longitude_of_ascending_node_range().midpoint()),
+              IsNear(54 * ArcSecond));
+  EXPECT_THAT(
+      AbsoluteError(*initial_osculating.argument_of_periapsis,
+                    elements.mean_argument_of_periapsis_range().midpoint()),
+      IsNear(61 * ArcSecond));
+
+  // Mean element stability.
+  EXPECT_THAT(elements.mean_semimajor_axis_range().measure(),
+              IsNear(1.0 * Milli(Metre)));
+  EXPECT_THAT(elements.mean_eccentricity_range().measure(), IsNear(1.0e-10));
+  EXPECT_THAT(elements.mean_inclination_range().measure(),
+              IsNear(0.61 * Micro(ArcSecond)));
+  EXPECT_THAT(elements.mean_longitude_of_ascending_node_range().measure(),
+              IsNear(1.9 * ArcMinute));
+  EXPECT_THAT(elements.mean_argument_of_periapsis_range().measure(),
+              IsNear(2.2 * ArcMinute));
+
+  OFStream f(SOLUTION_DIR / "mathematica" /
+             "unperturbed_elements.generated.wl");
+  f << mathematica::Assign("unperturbedOsculating",
+                           elements.osculating_equinoctial_elements());
+  f << mathematica::Assign("unperturbedMean",
+                           elements.mean_equinoctial_elements());
+}
+
+TEST_F(OrbitalElementsTest, J2Perturbation) {
+  Time const mission_duration = 10 * Day;
+
+  KeplerianElements<GCRS> initial_osculating;
+  initial_osculating.semimajor_axis = 7000 * Kilo(Metre);
+  initial_osculating.eccentricity = 1e-6;
+  initial_osculating.inclination = 10 * Milli(ArcSecond);
+  initial_osculating.longitude_of_ascending_node = 10 * Degree;
+  initial_osculating.argument_of_periapsis = 20 * Degree;
+  initial_osculating.mean_anomaly = 30 * Degree;
+  auto const status_or_elements = OrbitalElements::ForTrajectory(
+      *EarthCentredTrajectory(initial_osculating,
+                              J2000,
+                              J2000 + mission_duration,
+                              *oblate_earth_ephemeris_),
+      oblate_earth_,
+      MasslessBody{});
+  ASSERT_THAT(status_or_elements, IsOk());
+  OrbitalElements const& elements = status_or_elements.ValueOrDie();
+  EXPECT_THAT(elements.anomalistic_period() - *initial_osculating.period,
```

Comment:
Why `AbsoluteError` in the previous test and not here?

Reply:
In the previous test it is noise, I am testing that it is small; here it is due to the perturbation, its sign is interesting.